### DISCUSSION TITLE
[v0.23.0] Some changed specs and behaviours

### DISCUSSION THREAD

**Discussion by kaboc** [pocketbase/pocketbase#5675] [Q&A] [answered]
I started to examine how easy it is to migrate my app to v0.23.0-rc. I've already come across some difficulties, so I'd like to ask several questions.

## email

The `email` field in the users table was not "Noempty" before, but leaving it empty is no longer allowed. I'm using only OAuth2 at the moment. I'd like to hold users' personal information as little as possible, so tried to change the value in the field to an empty string, but it wasn't possible.

I guess the change is intended to improve safety in some way. Do you think it is safe to disable "Noempty" for myself? Can it cause trouble in some internal behaviour, in addition to issues related to safety/consistency?

## avatar

I'd like to delete the avatar data for the same reason as above. Unlike `email`, I also need to block an image from being stored or remove it after it is stored. The former is more preferable for me.

I tried `OnRecordAuthWithOAuth2Request()`, but could not modify the data before calling `e.Next()`, and have no idea how I can interfere with image storing. How can it be achieved?

## created, updated

Those fields used to be automatically updated even when the record was updated using a method of `app.DB()`, but v0.23.0-rc5 does not work the same way. Is the change intentional? I luckily have only a few places I need to change for this, but it may be much harder for large projects. 

(I haven't checked yet whether they are updated if records are changed not by `app.DB()` but by record operations equivalent to old Dao methods.)

### Before v0.23.0

```go
var externalAuths []models.ExternalAuth
err := db.
	Select().
	From("_externalAuths").
	Where(dbx.HashExp{"recordId": userId, "providerId": providerId}).
	All(&externalAuths)
```

This was working fine.

### v0.23.0-rc5

```go
var externalAuths []core.ExternalAuth
err := db.
	Select().
	From("_externalAuths").
	Where(dbx.HashExp{"recordRef": userId, "providerId": providerId}).
	All(&externalAuths)
```

This does not work now. (`externalAuths` becomes `[]`.) However, it works if the variable is non-slice and `One()` is used instead. This may be because of my Go skill issue, but can you/anyone tell me why `All()` does not work despite the same filtering?

The query below works although `All()` is used, which has made me even more confused.

```go
var records []*core.Record
err := app.RecordQuery("_externalAuths").
	Where(dbx.HashExp{"recordRef": userId, "providerId": providerId}).
	All(&records)
```

**Comment by ganigeorgiev** [comment, marked as answer]
1. Previously system fields like `id`, `email`, `emailVisibility`, etc. were not part of the collection fields definition and users didn't have control over them. Specifically for the email in v0.22 there IS also an option to make the email required but it is from the "Collection Options" tab. Now this is just migrated under the `email` field options. It is safe to disable the "Nonempty" toggle if the `email` is optional for your app.

2. I don't understand what do you mean. If you don't want the OAuth2 avatar to be assigned automatically to the user you can remove it from the optional mapping:
  ![oauth2_mapping](https://i.imgur.com/gF0brWP.png)

3. Again, it is not clear what do you mean. With v0.23.0 the `created` and `updated` fields are now part of the collection schema and can be removed if not needed.
    > when the record was updated using a method of app.DB()...
    
    Don't do this unless you have a good reason and understanding of how PocketBase work. Prefer operating with the `core.Record` and `app.Save(record)`.

4. `_externalAuths` is now a system collection and `core.ExternalAuth` is a record proxy.
    The below should work:
    ```go
    auths := []*core.ExternalAuth{}
    err := app.RecordQuery(core.CollectionNameExternalAuths).
        Where(dbx.HashExp{"recordRef": userId, "providerId": providerId}).
        All(&auths) // note: rarely you'll have multiple providers sharing the same id
    ```
    Note that if you want to retrieve all externalAuths of a specific auth record you can use:
    ```go
    auths, err := app.FindAllExternalAuthsByRecord(authRecord)
    ```

**Reply by kaboc** [reply]
Aha! What a nice improvement it is to allow for removal of unnecessary fields! Thank you for the advice about the other two. I got it.

> Don't do this unless you have a good reason and understanding of how PocketBase work. Prefer operating with the `core.Record` and `app.Save(record)`.

I understand we are not supposed to do it. It is also not good in that it does not trigger hooks and realtime notifications, either. However, most of operations in my PocketBase server are bulk upserts for offline sync, which I think are complex enough for me to resort to `dbx.Builder`.

I sometimes wish bulk inserts were supported. I need to loop through a data list to operate one by one. My concern about it is performance and load due to many operations instead of a single SQL executing them at once.

**Reply by ganigeorgiev** [reply]
> My concern about it is performance and load due to many operations instead of a single SQL executing them at once.

I can't help without a concrete example.

Keep in mind that with SQLite there is no "wire"/network call, the database lives as part of your app process and a query is more-or-less equivalent to calling a function (see also _["Many Small Queries Are Efficient In SQLite"](https://sqlite.org/np1queryprob.html)_).

If you want your inserts to be "all or nothing" then you could wrap them in a transaction.

**Reply by kaboc** [reply]
In my understanding, a for-loop of record operations produces many `INSERT` statements, whether they are wrapped in a transaction. Despite the high performance of SQLite and its mechanism that is very different from other RDMBS, bulk inserting with `INSERT INTO ... VALUES (...), (...), (...), (...), ...` (which I do instead) is better performance-wise in my opinion. I'd like to avoid more than 1k insert statements in a row. (I think the "200 or more SQL statement per webpage" written in the linked page is mainly about `SELECT`.)

But, never mind! PocketBase is already great without it.

**Reply by ganigeorgiev** [reply]
> bulk inserting with INSERT INTO ... VALUES (...), (...), (...), (...), ... (which I do instead) is better performance-wise in my opinion.

A single write transaction wrapping all inserts should produce similar execution time to a single large INSERT statement.

I've tried it with the logs bulk insertion and there wasn't significant difference (see the [related code comment](https://github.com/pocketbase/pocketbase/blob/master/core/base.go#L1240-L1241)).

I'll stop engaging as I find performance discussions pointless if there is nothing concrete to analyze or work on.
